File: 1036-rotting-oranges/1036-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        Queue<Pair> queue = new LinkedList<>();
        int fresh = 0;
        int time = 0;

        int m = grid.length;
        int n = grid[0].length;

        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(grid[i][j]==2){
                    queue.add(new Pair<>(i,j));
                }
                else if(grid[i][j]==1){
                    fresh++;
                }
            }
        }

        int[][] directions = {{1,0}, {-1,0}, {0,-1}, {0,1}};

        while(!queue.isEmpty() && fresh>0){
            int len = queue.size();

            for(int i=0; i<len; i++){
                Pair<Integer, Integer> cord = queue.poll();
                int x = cord.getKey();
                int y = cord.getValue();

                for(int[] dir: directions){
                    int xn = x + dir[0];
                    int yn = y + dir[1];

                    if(xn<0 || xn>=m || 
                        yn<0 || yn>=n){
                            continue;
                        }

                    if(grid[xn][yn]==1){
                        fresh--;
                        grid[xn][yn]=2;
                        queue.add(new Pair<>(xn, yn));
                    }
                }
            }
            time++;

        }

        if(fresh==0){
            return time;
        }
        return -1;

    }
}

File: 1036-rotting-oranges/1036-rotting-oranges.py
class Solution(object):
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])
        q = deque()
        n_oranges = 0

        for i in range(0, m):
            for j in range(0, n):
                if grid[i][j] == 2:
                    q.append((i,j))
                if grid[i][j] == 1:
                    n_oranges += 1
                
        
        if n_oranges == 0:
            return 0
            
        count = 0
        rotten = 0

        while q:
            size = len(q)
            count+=1

            for _ in range(0, size):
                i, j = q.popleft()

                for nx, ny in [(0,1), (0,-1), (1,0), (-1,0)]:
                    nr = i + nx
                    nc = j + ny

                    if nr < 0 or nr >= m or nc < 0 or nc >= n or grid[nr][nc] == 0:
                        continue

                    if grid[nr][nc] == 1:
                        rotten+=1
                        grid[nr][nc] = 2
                        q.append((nr,nc))

        if rotten==n_oranges:
            return count-1
        return -1